### ISSUE TITLE
Grpc Load Balancing not working

### ISSUE THREAD

**Issue by amanullah92** [aws/aws-app-mesh-controller-for-k8s#427]
**Describe the bug**
GRPC Load Balancing is not working. VirtualRouter is sending all requests to only one pod.

**Steps to reproduce**
Create two services (let's call them A and B), with one service (A) receiving traffic from outside world, and invoking grpc method on the other service (B). Create multiple pods for service B.

**Expected outcome**
Traffic should be equally distributed between pods of service B.

**Environment**

* App Mesh controller version: appmesh-controller:v1.1.1
* Envoy version: aws-appmesh-envoy:v1.15.1.0-prod
* Are you using any integrations? X-ray, Jaeger etc. If so versions? No
* Kubernetes version: 1.18
* Using EKS (yes/no), if so version? yes, 1.18

**Additional Context**:
Here are the configurations:
`
apiVersion: appmesh.k8s.aws/v1beta2
kind: VirtualNode
metadata:
  name: grpc-demo-service
  namespace: amaan
spec:
  awsName: grpc-demo-service
  podSelector:
    matchLabels:
      app: grpc-demo-service
  listeners:
    - portMapping:
        port: 9998
        protocol: grpc
  logging:
    accessLog:
      file:
        path: "/dev/stdout"
  serviceDiscovery:
    dns:
      hostname: grpc-demo-service.amaan.svc.cluster.local
---
apiVersion: appmesh.k8s.aws/v1beta2
kind: VirtualService
metadata:
  name: grpc-demo-service
  namespace: amaan
spec:
  awsName: grpc-demo-service
  provider:
    virtualRouter:
      virtualRouterRef:
        name: grpc-demo-service

---
apiVersion: appmesh.k8s.aws/v1beta2
kind: VirtualRouter
metadata:
  namespace: amaan
  name: grpc-demo-service
spec:
  listeners:
    - portMapping:
        port: 80
        protocol: grpc
  routes:
    - name: default_route
      grpcRoute:
        action:
          weightedTargets:
            - virtualNodeRef:
                name: grpc-demo-service
              weight: 1
----
apiVersion: v1
kind: Service
metadata:
  name: grpc-demo-service
  namespace: amaan
spec:
  ports:
    - port: 9998
      targetPort: 9998
  selector:
    app: grpc-demo-service

`

**Comment by lavignes** [comment]
Hi @amanullah92 could you possibly be seeing the behavior described in #70 ?

App Mesh does not configure nor expose the option to enable true round-robin load-balancing across all discovered endpoints today, but instead uses Envoy's default behavior. This behavior can be beneficial in many cases, only creating more TCP connections when needed to handle more traffic, but is obviously not necessity intuitive.

This might be a dup of #237 if I am right.

**Comment by amanullah92** [comment]
I'm sorry, if I get correctly those issues are in context of TLS. I'm not using TLS between the services as of now.

_App Mesh does not configure nor expose the option to enable true round-robin load-balancing across all discovered endpoints today, but instead uses Envoy's default behavior. This behavior can be beneficial in many cases, only creating more TCP connections when needed to handle more traffic, but is obviously not necessity intuitive._

I can see from sidecar Config Dump that it has discovered only one pod, and all stats are for that pod.  The other pod is completely missing. So I'm just wondering if it's just a Envoy load balancing issue or a service discovery issue. I've tried with headless as non-headless Kubernetes service- in both cases traffic distribution is not taking place.


cds_egress_custom-mesh_grpc-demo-service_grpc_9998::default_priority::max_connections::2147483647
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::default_priority::max_pending_requests::2147483647
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::default_priority::max_requests::2147483647
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::default_priority::max_retries::2147483647
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::high_priority::max_connections::1024
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::high_priority::max_pending_requests::1024
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::high_priority::max_requests::1024
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::high_priority::max_retries::3
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::added_via_api::true
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::**10.0.1.215:9998**::cx_active::2
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::cx_connect_fail::0
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::cx_total::2
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::rq_active::9
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::rq_error::0
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::rq_success::12987
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::rq_timeout::0
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::rq_total::12996
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::hostname::grpc-demo-service.amaan.svc.cluster.local
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::health_flags::healthy
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::weight::1
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::region::
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::zone::
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::sub_zone::
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::canary::false
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::priority::0
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::success_rate::-1.0
cds_egress_custom-mesh_grpc-demo-service_grpc_9998::10.0.1.215:9998::local_origin_success_rate::-1.0

**Comment by lavignes** [comment]
Oh @amanullah92 sorry about that. The links to the issues I posted above were wrong. :/ I meant to link items from the App Mesh roadmap.

**Comment by amanullah92** [comment]
Thanks.

Is it normal that I am not seeing both end points in the envoy side car config? I can see that Envoy has 3 active connections to that pod at the time of load (150-200 TPS). It's uncanny that all three connections chose the same pod.

**Comment by vladi-algo** [comment]
@amanullah92 - Have you figured out the reason of this behaviour?

**Comment by sunao-uehara** [comment]
Same issue happens to me as well. It seems it is expected behavior as of now? I knew that gRPC connection is sticky, but I thought envoy proxy via AppMesh solve this issue.

**Comment by amanullah92** [comment]
The reason for this behaviour is that in the default configuration of Envoy provided by AWS, the DNS resolution mechanism used is Logical DNS. In this case, for every new connection created by Envoy, a DNS query is made for the backend service endpoint and the first returned IP is used. Now, when you use GRPC, it uses http2, which means as few as few one connection for all requests. Which means, all traffic ends up on a single host in the backend. 

We ended up using CloudMap service discovery mechanism with AppMesh, which did not have this problem: In this case, App mesh configures Envoy with individual IPs of pods/servers returned by the cloud map service, thus allowing envoy to truly load balance GRPC/Http2 requests, instead of load balancing connections (which works okay for http1.1, since connections are short lived).

You can read more about it here:
https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/service_discovery#arch-overview-service-discovery-types-logical-dns

**Comment by dastbe** [comment]
Thanks for the writeup @amanullah92 . To give some context, the default for App Mesh to use LOGICAL_DNS (think of this as "these endpoints are loadbalancers") was to best support R53. R53 by default will only return up to 8 entries, and after that will randomize the response. If we had used STATIC_DNS ( think of this as "these endpoints are endpoints") then Envoy would've constantly churned connections and led to an even worse experience.

That being said, this is not a limitation of CoreDNS and shouldn't be a limitation we impose on you. We recently added support for configuring the dns response type in App Mesh, and it's been [exposed in the controller](https://github.com/aws/aws-app-mesh-controller-for-k8s/commit/56c8736321a3c1c076ba56574f9c2b4a6619603e). Setting this will make load balancing "just work" with the existing DNS available in your cluster.

We also look into making this the default when using the k8s controller (because it's what you want).
